exonum-java-binding-common/src/main/java/com/exonum/binding/common/message/BinaryTransactionMessage.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.exonum.binding.common.message;
+
+import static com.exonum.binding.common.hash.Hashing.sha256;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.nio.ByteOrder.LITTLE_ENDIAN;
+import static java.util.Arrays.copyOf;
+
+import com.exonum.binding.common.crypto.PublicKey;
+import com.exonum.binding.common.hash.HashCode;
+import com.google.common.base.Objects;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+/**
+ * Binary implementation of the {@link TransactionMessage} class. Immutable by design.
+ */
+public final class BinaryTransactionMessage implements TransactionMessage {
+
+  private final ByteBuffer rawTransaction;
+
+  BinaryTransactionMessage(byte[] bytes) {
+    this(ByteBuffer.wrap(bytes));
+  }
+
+  BinaryTransactionMessage(ByteBuffer buffer) {
+    ByteBuffer slice = buffer.slice();
+    checkArgument(MIN_MESSAGE_SIZE <= slice.limit(),
+        "Transaction message requires at least %s bytes space, but was %s",
+        MIN_MESSAGE_SIZE, slice.limit());
+
+    this.rawTransaction = ByteBuffer.allocate(slice.limit()).order(LITTLE_ENDIAN);
+    this.rawTransaction.put(slice);
+  }
+
+  @Override
+  public PublicKey getAuthor() {
+    byte[] key = new byte[AUTHOR_PUBLIC_KEY_SIZE];
+    rawTransaction.position(AUTHOR_PUBLIC_KEY_OFFSET);
+    rawTransaction.get(key);
+    return PublicKey.fromBytes(key);
+  }
+
+  @Override
+  public short getServiceId() {
+    return rawTransaction.getShort(SERVICE_ID_OFFSET);
+  }
+
+  @Override
+  public short getTransactionId() {
+    return rawTransaction.getShort(TRANSACTION_ID_OFFSET);
+  }
+
+  @Override
+  public byte[] getPayload() {
+    int payloadSize = rawTransaction.limit() - MIN_MESSAGE_SIZE;
+    byte[] payload = new byte[payloadSize];
+    rawTransaction.position(PAYLOAD_OFFSET);
+    rawTransaction.get(payload);
+    return payload;
+  }
+
+  @Override
+  public HashCode hash() {
+    return sha256().newHasher()
+        .putBytes(rawTransaction.duplicate())
```

Comment:
It is broken as the code that gets byte arrays modifies position of the original buffer, therefore, the bytes that get hashed depend on the invocations of `getPayload`/`getSignature`/`getAuthor`. 

Can we fix it so that the implementation is more reliable? E.g., as the bb is internally allocated and guaranteed to have an array corresponding to the full message, hash the array (with a comment clarifying this implementation)?